docs/docs/reference/custom-components/02-custom-component.md

```diff
@@ -0,0 +1,647 @@
+
+The validation function can freely mutate the incoming parameter value before returning true, for example to constrain
+a numerical value between some upper and lower bounds.
+
+<Tabs groupId="programming-language">
+<TabItem value="python" label="Python">
+
+```python title="custom_component.py"
+    def on_validate_parameter_callback(self, parameter: sr.Parameter) -> bool:
+        if parameter.get_name() == "A":
+            if parameter.is_empty():
+                self.get_logger().warn("Parameter A cannot be empty")
+                return False
+        elif parameter.get_name() == "B":
+            if parameter and parameter.get_value() < 0.0:
+                self.get_logger().warn(f"Parameter B cannot be negative ({parmaeter.get_value()}). "
+                                        "Setting value to 0.0 instead")
+                parameter.set_value(0.0)
+        return True
+```
+
+</TabItem>
+<TabItem value="c++" label="C++">
+
+```cpp title="include/custom_component_package/CppComponent.hpp"
+class CppComponent : public modulo_components::Component {
+// ...
+
+protected:
+  bool
+  on_validate_parameter_callback(const std::shared_ptr<state_representation::ParameterInterface>& parameter) override;
+};
+```
+
+```cpp title="src/CppComponent.cpp"
+bool CppComponent::on_validate_parameter_callback(
+    const std::shared_ptr<ParameterInterface>& parameter
+) {
+  if (parameter->get_name() == "A") {
+    if (parameter->is_empty()) {
+      RCLCPP_WARN(this->get_logger(), "Parameter A cannot be empty");
+      return false;
+    }
+  } else if (parameter->get_name() == "B") {
+    if (parameter && parameter->get_value() < 0.0) {
+      RCLCPP_WARN(this->get_logger(), "Parameter B cannot be negative (%f). Setting value to 0.0 instead",
+                  parameter->get_value());
+      parameter->set_value(0.0);
+    }
+  }
+  return true;
+}
+```
+
+</TabItem>
+</Tabs>
+
+
+:::tip
+
+Use the logging interface provided by `modulo_components` wherever appropriate. This will forward log messages from
+custom components to ROS logs and the user interface.
+
+:::
+
+## Adding signals
+
+Components make it possible to bind data objects to publishers or subscribers with class attributes. For inputs, this
+means that the associated input data object is automatically updated every time a new message is received. For outputs,
+this means that internal changes to the output data object will automatically be updated in the publisher.
+
+The binding logic is supported for a number of common message data types. To interface with non-standard messages
+(for example, to communicate with ROS 2 nodes using custom message types outside the AICA framework), it is possible to
+define raw publishers and subscribers following standard ROS 2 conventions.
+
+### Message types
+
+The supported message types are defined in the `std_msgs` library and include:
+
+- Boolean
```

Comment:
Is this capital B intended?

Reply:
Yes, it's like Cartesian, named after some old dude.